### PR TITLE
Add support for github labels ignore list

### PR DESCRIPTION
**Pull request by udisch**
Adds the option to configure github labels which, if present on the pull request, will cause the build to be skipped. This is useful, for example, when working on a pull request that is still in progress, and you don't want commits to trigger a build.

https://github.com/jenkinsci/ghprb-plugin/issues/210

### CODE REVIEW

**Comment by byaminov** [comment]
👍

**Comment by angel-git** [comment]
👍 👍

**Comment by benpatterson** [comment]
Udi - I'm planning to give this a look in the coming days.

**Comment by benpatterson** [comment]
Hi @udisch overall looks pretty good. Could you please make a couple of modifications?

* Can you change `labelslist` to `labelsIgnoreList` or something along those lines? Basically I'd like to leave room for creating a labels whitelist later.
* I see you've updated tests which is great, but I'd like a test that asserts a build is/is not triggered depending on the label that was found on the pull request. I didn't see a test for that but it would help prevent regressions later.

Thanks for the pull request!

**Comment by udisch** [comment]
@benpatterson Thanks for the feedback!

**Comment by udisch** [comment]
@benpatterson I added a test and renamed the labels field. Also fixed some small bugs in my PR
which I found while testing:
1) Labels weren't checked when adding comments
2) Labels read from the jenkins config assumed whitespace as the separator, but github allows spaces labels, so now newline is used.

Also note that I didn't add any handling for hooks - Removing a label won't trigger anything, but it's possible to use "retest this please".

Thanks!

File: src/main/resources/org/jenkinsci/plugins/ghprb/GhprbTrigger/global.groovy [hunk 1]
```diff
@@ -46,7 +46,7 @@ f.section(title: descriptor.displayName) {
     f.entry(field: "cron", title: _("Crontab line"), help: "/descriptor/hudson.triggers.TimerTrigger/help/spec") {
       f.textbox(default: "H/5 * * * *", checkUrl: "'descriptorByName/hudson.triggers.TimerTrigger/checkSpec?value=' + encodeURIComponent(this.value)") 
     }
-    f.entry(field: "labelslist", title: _("List of github labels for which the build should not be triggered.")) {
+    f.entry(field: "labelsIgnoreList", title: _("List of github labels for which the build should not be triggered.")) {
```
Comment (benpatterson) [inline src/main/resources/org/jenkinsci/plugins/ghprb/GhprbTrigger/global.groovy]: As per your point on the fix you implemented, could you point out that this list needs to have carriage return characters as separators?

**Comment by udisch** [comment]
Updated.

**Comment by benpatterson** [comment]
thanks @udisch 

Psyched

**Comment by udisch** [comment]
@benpatterson Do you know when the next plugin release is due?

**Comment by benpatterson** [comment]
Hi @udisch we have released this now. :)

**Comment by udisch** [comment]
@benpatterson Thanks! I don't see it in the jenkins plugin repository yet..

**Comment by jaggernod** [comment]
@benpatterson Thank you for the great feature. The version is already out but I'm not able to use it as GhprbTriggerContext doesn't support it yet. Any ideas when would that happen? Thanks.

**Comment by benpatterson** [comment]
Unfortunately it looks like this has caused #441. I'm planning on reverting unless I get a workaround to the underlying issue in the github-api plugin.

**Comment by benpatterson** [comment]
@udisch Unfortunately this change has caused a regression in functionality. That's described in the thread for issue 441. I'm reverting this change in order to fix the plugin, then we can look at label options after that, likely in tandem with another contributor.

**Comment by udisch** [comment]
@benpatterson Thanks, and sorry. What happened exactly? Was the api removed from the github-api lib?

**Comment by benpatterson** [comment]
Hi @udisch label support is still available in the github-api library; however, it turns out that when calling getLabels, it uses GitHubs issues API instead of its pull request API. When making that call, it rewrites the `pr` object, including the root url to use for other calls (e.g., .../issues/... in the url, instead of .../pulls/...). Any subsequent call by the plugin would return a 404 because those calls use the pull request API endpoints.

If you want, you can walk through it on the issue 441 thread. I really like your feature and would like to see it re-implemented. There is another PR that you'll see was referenced. I'll do some rebasing, etc, and CC you on it when I'm done. From there perhaps we can tackle the underlying issue. (Or we will benefit from a fix on the underlying github-api  😎 )

**Comment by udisch** [comment]
On the github api, labels are only available on the /issues endpoint: https://developer.github.com/v3/pulls/#labels-assignees-and-milestones

**Comment by benpatterson** [comment]
@udisch Yes... I'm not arguing that we use some other API. The issue is that the underlying implementation on the other plugin ends up re-writing the pr object as an issue object (basically).

I think re-implementing this would require a workaround to that problem...unless we get more details on the other issue. I'm not sure why it was implemented the way it was....it could be a  bug, or it could be by design.

**Pull request merged**